### PR TITLE
ESP32: Add optional NimBLE support for BLE transport

### LINKED ISSUE #1274: Replace Bluedroid with NimBLE
> Is the 300+ KB increase on .text and +79KB .data increase for the ESP32 examples really correct? OUCH!

That's likely due to Bluedroid as BT got pulled in. I am working on a NimBLE implementation for the BT layer, that should help bring this down. Hopefully will raise a PR early next week.

_Originally posted by @kedars in https://github.com/project-chip/connectedhomeip/pull/1241#issuecomment-649584806_

### PR DESCRIPTION
**Pull request by prasad-alatkar**
#### Problem

This PR adds NimBLE as optional BLE transport layer for ESP32 platform. I have verified rendezvous session establishment with chip-tool example.

<the issue this PR is intended to address>

 #### Summary of Changes

- Adds `BLEManagerImpl.cpp` and `BLEManagerImpl.h` for NimBLE.
- Refactor ESP32 into `bluedroid` and `nimble` sub-directories with corresponding `BLEManagerImpl.cpp` files.
- Modified ~~`src/platform/Makefile.am` and~~ `src/platform/BUILD.gn` to accommodate `nimble` support in ~~make and~~ GN build system. 

fixes https://github.com/project-chip/connectedhomeip/issues/1274 and https://github.com/project-chip/connectedhomeip/issues/1115

### CODE REVIEW

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,947 @@
...
+
+/**
+ *    @file
+ *          Provides an implementation of the BLEManager singleton object
+ *          for the ESP32 (NimBLE) platform.
+ */
+/* this file behaves like a config.h, comes first */
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+
+#if CHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE
```
Comment (andy31415) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:27]: What are the thoughts about removing the entire file from compilation from a compile setting rather than checking #ifdefs like this on the entire file?

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+
+const uint8_t CharProps_ReadNotify = BLE_GATT_CHR_F_READ | BLE_GATT_CHR_F_NOTIFY;
+const uint8_t CharProps_Write      = BLE_GATT_CHR_F_WRITE;
+
+/*XXX TODO*/
+
+}//unnamed namespace
+
+
+/*XXX TODO*/
```
Comment (andy31415) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:92]: github issue here and above?

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,947 @@
...
+    },
+};
+
+
+CHIP_ERROR BLEManagerImpl::_Init()
+{
+    CHIP_ERROR err;
+
+    mServiceMode = ConnectivityManager::kCHIPoBLEServiceMode_Enabled;
+    /*XXX for (int i = 0; i < kMaxConnections; i++)
```
Comment (andy31415) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:130]: TODO here rather than commented out code

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,947 @@
...
+    }
+    else
+    {
+        mDeviceName[0] = 0;
+        ClearFlag(mFlags, kFlag_UseCustomDeviceName);
+    }
+    return CHIP_NO_ERROR;
+
+exit:
+    return CHIP_ERROR_INCORRECT_STATE;
```
Comment (andy31415) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:235]: Should we preserve 'err' and have

```
exit:
    return err;
```

instead? here we lose the err from the call on line 225 and we get a generic 'incorrect state' which loses some error info.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,947 @@
...
+    }
+    strcpy(buf, mDeviceName);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BLEManagerImpl::_SetDeviceName(const char * deviceName)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (mServiceMode == ConnectivityManager::kCHIPoBLEServiceMode_NotSupported)
```
Comment (andy31415) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:213]: VerifyOrExit macro could be used to have fewer return points in this function.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 6]
```diff
@@ -0,0 +1,947 @@
...
+
+    for (uint16_t i = 0; i < kMaxConnections; i++)
+    {
+        if (mSubscribedConIds[i] == conId)
+        {
+            return CHIP_NO_ERROR;
+        }
+        else if (mSubscribedConIds[i] == BLE_CONNECTION_UNINITIALIZED && i < freeIndex)
+        {
+            freeIndex = i;
```
Comment (andy31415) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:765]: should we have a break here ?

**Comment by woody-apple** [comment]
Given this is WIP, moving to Draft

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 2 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: Removed this TODO.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 3 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: Removed, Thanks.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 4 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: Have replaced with `err`.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 5 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: Done !! Thank you

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 6 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp:747]: Yes, thank you.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 7]
```diff
@@ -0,0 +1,947 @@
...
+const ChipBleUUID chipUUID_CHIPoBLEChar_RX = { { 0x18, 0xEE, 0x2E, 0xF5, 0x26, 0x3D, 0x45, 0x59, 0x95, 0x9F, 0x4F, 0x9C, 0x42, 0x9F,
+                                                 0x9D, 0x11 } };
+ble_uuid128_t UUID_CHIPoBLEChar_RX = { { BLE_UUID_TYPE_128 } , { 0x18, 0xEE, 0x2E, 0xF5, 0x26, 0x3D, 0x45, 0x59, 0x95, 0x9F, 0x4F, 0x9C, 0x42, 0x9F,
+                                                 0x9D, 0x11 } };
+
+const ble_uuid128_t UUID128_CHIPoBLEChar_TX = { BLE_UUID_TYPE_128, { 0x12, 0x9D, 0x9F, 0x42, 0x9C, 0x4F, 0x9F, 0x95, 0x59, 0x45, 0x3D, 0x26, 0xF5,
+                                                  0x2E, 0xEE, 0x18 } };
+const ChipBleUUID chipUUID_CHIPoBLEChar_TX = { { 0x18, 0xEE, 0x2E, 0xF5, 0x26, 0x3D, 0x45, 0x59, 0x95, 0x9F, 0x4F, 0x9C, 0x42, 0x9F,
+                                                 0x9D, 0x12 } };
+ble_uuid128_t UUID_CHIPoBLEChar_TX = { { BLE_UUID_TYPE_128  }, { 0x12, 0x9D, 0x9F, 0x42, 0x9C, 0x4F, 0x9F, 0x95, 0x59, 0x45, 0x3D, 0x26, 0xF5,
```
[the remark this answers is not in the document]
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: Done!! Thanks.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 8]
```diff
@@ -0,0 +1,947 @@
...
+};
+
+const uint16_t UUID16_CHIPoBLEService = 0xFEAF;
+const ble_uuid16_t UUID_CHIPoBLEService = { BLE_UUID_TYPE_16, UUID16_CHIPoBLEService };
+
+const ble_uuid128_t UUID128_CHIPoBLEChar_RX = { BLE_UUID_TYPE_128, { 0x11, 0x9D, 0x9F, 0x42, 0x9C, 0x4F, 0x9F, 0x95, 0x59, 0x45, 0x3D, 0x26, 0xF5,
+                                                  0x2E, 0xEE, 0x18 } };
+const ChipBleUUID chipUUID_CHIPoBLEChar_RX = { { 0x18, 0xEE, 0x2E, 0xF5, 0x26, 0x3D, 0x45, 0x59, 0x95, 0x9F, 0x4F, 0x9C, 0x42, 0x9F,
+                                                 0x9D, 0x11 } };
+ble_uuid128_t UUID_CHIPoBLEChar_RX = { { BLE_UUID_TYPE_128 } , { 0x18, 0xEE, 0x2E, 0xF5, 0x26, 0x3D, 0x45, 0x59, 0x95, 0x9F, 0x4F, 0x9C, 0x42, 0x9F,
```
[the remark this answers is not in the document]
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: Done, Thanks.

**Comment by prasad-alatkar** [comment]
How can I validate the implementation using rendezvous ? Is there any documentation available for rendezvous ?

**Comment by kedars** [comment]
> How can I validate the implementation using rendezvous ? Is there any documentation available for rendezvous ?

cc @vivien-apple @shana-apple

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 9]
```diff
@@ -0,0 +1,926 @@
...
+            }
+        }
+    }
+    return false;
+}
+
+int BLEManagerImpl::ble_svr_gap_event(struct ble_gap_event * event, void * arg)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    // BLEManagerImpl BLEGap;
```
Comment (vivien-apple) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:794]: Remove the commented line.

File: src/platform/Makefile.am [hunk 10]
```diff
@@ -107,6 +107,7 @@ noinst_HEADERS                          = \
     @top_srcdir@/src/platform/ESP32/CHIPDevicePlatformConfig.h \
     @top_srcdir@/src/platform/ESP32/ConnectivityManagerImpl.h \
     @top_srcdir@/src/platform/ESP32/BLEManagerImpl.h \
+    @top_srcdir@/src/platform/ESP32/nimble/BLEManagerImpl.h \
```
Comment (vivien-apple) [inline src/platform/Makefile.am:110]: Can we add src/platform/bluedroid directory too and move the current BLEManagerImpl.h and BLEManagerImpl.cpp there ?

Also can we use CONFIG_BT_BLUEDROID_ENABLED / CONFIG_BT_NIMBLE_ENABLED to conditionally add those files to the build 
There reason I'm asking is because in the current shape even if I enabled NimBLE in menuconfig it stills try to build the current BLEManagerImpl.h/cpp files related to Bluedroid.

File: src/platform/Makefile.am [hunk 11]
```diff
@@ -347,6 +348,7 @@ if CHIP_DEVICE_LAYER_TARGET_ESP32
 
 libDeviceLayer_a_SOURCES               +=       \
     ESP32/BLEManagerImpl.cpp                    \
+    ESP32/nimble/BLEManagerImpl.cpp             \
```
Comment (vivien-apple) [inline src/platform/Makefile.am:351]: See previous comment.

File: src/platform/ESP32/nimble/BLEManagerImpl.h [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+
+/**
+ *    @file
+ *          Provides an implementation of the BLEManager singleton object
+ *          for the ESP32 platform (NimBLE).
+ */
+
+#ifndef BLE_MANAGER_IMPL_H
+#define BLE_MANAGER_IMPL_H
+
```
Comment (vivien-apple) [inline src/platform/ESP32/nimble/BLEManagerImpl.h:27]: I'm not a huge fan of #ifdef in implementation files as it makes code hard to follow. But since this file is just a header and since the only differences with src/platform/ESP32/BLEManagerImpl.h are #includes method calls related to BLE events should we use #ifdef instead of adding a new file ?

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 13]
```diff
@@ -0,0 +1,926 @@
...
+        PlatformMgr().PostEvent(&event);
+        buf = NULL;
+    }
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(DeviceLayer, "HandleRXCharWrite() failed: %s", ErrorStr(err));
+    }
+    PacketBuffer::Free(buf);
```
Comment (vivien-apple) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:576]: Should we call PacketBuffer::Free if ```buf == NULL``` ?

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 14]
```diff
@@ -0,0 +1,926 @@
...
+CHIP_ERROR BLEManagerImpl::ConfigureAdvertisingData(void)
+{
+    CHIP_ERROR err;
+    struct ble_hs_adv_fields fields = { 0 }, scan_rsp = { 0 };
+
+    ESP32ChipServiceData chipServiceData;
+
+    // If a custom device name has not been specified, generate a CHIP-standard name based on the
+    // bottom digits of the Chip device id.
+    const uint16_t discriminator = 0x0F00;
```
Comment (vivien-apple) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:500]: Can we retrieve the discriminator from Config and use digits instead of hex ? 
```
    // If a custom device name has not been specified, generate a CHIP-standard name based on the
    // discriminator value
    uint32_t discriminator;
    SuccessOrExit(err = ConfigurationMgr().GetSetupDiscriminator(discriminator));

    if (!GetFlag(mFlags, kFlag_UseCustomDeviceName))
    {
        snprintf(mDeviceName, sizeof(mDeviceName), "%s%04u", CHIP_DEVICE_CONFIG_BLE_DEVICE_NAME_PREFIX, discriminator);
        mDeviceName[kMaxDeviceNameLength] = 0;
    }
```

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 15]
```diff
@@ -0,0 +1,926 @@
...
+    err = ConfigurationMgr().GetBLEDeviceIdentificationInfo(chipServiceData.DeviceIdInfo);
+    SuccessOrExit(err);
+
+    scan_rsp.tx_pwr_lvl_is_present = 1;
+    scan_rsp.tx_pwr_lvl            = BLE_HS_ADV_TX_PWR_LVL_AUTO;
+    scan_rsp.svc_data_uuid16       = (uint8_t *) UUID16_CHIPoBLEService;
+    scan_rsp.svc_data_uuid16_len   = sizeof(UUID16_CHIPoBLEService);
+
+    ble_gap_adv_set_fields(&fields);
+    ble_gap_adv_rsp_set_fields(&scan_rsp);
```
Comment (vivien-apple) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:537]: I tried to change the advertising data to look likes what the spec says. By any chance can you use raw data here ? That makes it easier to read imho. See https://github.com/project-chip/connectedhomeip/blob/master/src/platform/ESP32/BLEManagerImpl.cpp#L650

Also for the moment rendezvous expects the advertisement data to looks like the given link

**Comment by vivien-apple** [comment]
> > How can I validate the implementation using rendezvous ? Is there any documentation available for rendezvous ?
> 
> cc @vivien-apple @shana-apple

Assuming the ESP32 is configured and flashed properly, you need to run ```out/debug/standalone/chip-standalone-demo echo-ble 3840 1234567``` from your macOS build.

Basically 3840 correspond to 0x0F00 (default discriminator) and 1234567 is the default PIN Code used for testing.

I will try to update the documentation with complete build steps ASAP.
In the meantime feel free to ping me if you have any questions.

File: src/platform/ESP32/BLEManagerImpl.h [hunk 16]
```diff
@@ -205,6 +209,210 @@ inline bool BLEManagerImpl::_IsAdvertising(void)
 } // namespace DeviceLayer
 } // namespace chip
 
+#elif CONFIG_BT_NIMBLE_ENABLED
```
Comment (turon) [inline src/platform/ESP32/BLEManagerImpl.h:212]: Would it be possible to split the various ports into their own headers?

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 1 above)
Comment (turon) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp:27]: We need to decide which model we'd like to follow.  I had proposed this pattern of #ifdef wrapped sources, which is what OpenThread uses, so we avoid build system-specific dependencies and can configure the entire project with a simple config.h.

Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp:27]: Hi @andy31415 @turon sorry for delayed response, I have used "#ifdef" with this project. Please let me know if you have any better alternatives in mind.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 14 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: @vivien-apple yes the suggested changes are included

**Comment by prasad-alatkar** [comment]
@vivien-apple @turon @andy31415 @mspang I have modified the PR, please take a fresh look. I have verified the rendezvous session establishment using chip-tool. Thanks.

File: src/platform/ESP32/BLEManagerImpl.h (hunk 16 above)
Comment (vivien-apple) [reply src/platform/ESP32/BLEManagerImpl.h:212]: Sorry if I was unclear in the previous commit. I was more thinking of having the same BLEManagerImpl definition with an #ifdef for the platform specific methods. See commit https://github.com/vivien-apple/connectedhomeip-1/commit/9c728e07003f3fc9a5d77259b529e4ada721012b that goes on top of your PR :)

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 17]
```diff
@@ -0,0 +1,1074 @@
...
+    }
+    if (buf != NULL)
+    {
+        PacketBuffer::Free(buf);
+    }
+}
+
+void BLEManagerImpl::HandleTXCharRead(struct ble_gatt_char_context * param)
+{
+    /* XXX Not supported */
```
Comment (vivien-apple) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:693]: Does the XXX means it will be supported at some point ? Otherwise if that's something we don't plan to support we can get rid of the XXX.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 18]
```diff
@@ -0,0 +1,1074 @@
...
+    ChipLogError(DeviceLayer, "BLEManagerImpl::HandleTXCharRead() not supported");
+}
+
+void BLEManagerImpl::HandleTXCharCCCDRead(void * param)
+{
+    /* Not Supported XXX */
```
Comment (vivien-apple) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:699]: See previous comment.

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp [hunk 19]
```diff
@@ -0,0 +1,1074 @@
...
+        }
+    }
+
+    else
+    {
+        // If indications had previously been enabled for this connection, record that they are no longer
+        // enabled and inform the BLE layer that the client has "unsubscribed" the connection.
+        if (UnsetSubscribed(gapEvent->subscribe.conn_handle))
+        {
+            // HandleUnsubscribeReceived(gapEvent->subscribe.conn_handle, &CHIP_BLE_SVC_ID, &chipUUID_CHIPoBLEChar_TX);
```
Comment (vivien-apple) [inline src/platform/ESP32/nimble/BLEManagerImpl.cpp:737]: Is it commented on purpose ?

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 17 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: Done!!

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 18 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: Done!!

File: src/platform/ESP32/nimble/BLEManagerImpl.cpp (hunk 19 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/nimble/BLEManagerImpl.cpp]: Yes, will remove this unnecessary line.

File: src/platform/ESP32/BLEManagerImpl.h [hunk 20]
```diff
@@ -28,9 +28,39 @@
...
+#include "console/console.h"
+#include "esp_nimble_hci.h"
+#include "host/ble_hs.h"
+#include "host/util/util.h"
+#include "nimble/nimble_port.h"
+#include "nimble/nimble_port_freertos.h"
+#include "services/gap/ble_svc_gap.h"
+#undef min
+#undef max
+
```
Comment (vivien-apple) [inline src/platform/ESP32/BLEManagerImpl.h:53]: It seems like there is a lot of duplicates #include between this file and nimble/BLEManagerImpl.cpp. Is there any includes that can be removed from this .h file ?

Comment (prasad-alatkar) [reply src/platform/ESP32/BLEManagerImpl.h:47]: Thank you @vivien-apple , removed the unnecessary ones !!

File: src/platform/ESP32/BLEManagerImpl.h (hunk 16 above)
Comment (prasad-alatkar) [reply src/platform/ESP32/BLEManagerImpl.h:189]: Hi @vivien-apple I have made the relevant changes !! Thanks.

**Review by rwalker-apple — DISMISSED** [review]
the Dockerfile update implies a version change.  can you please split the dockerfile update into a separate PR (and include an update to integrations/docker/images/chip-build/version)?

The Dockerfile PR would need to land first, and a docker.io push will be necessary.  Then this PR would have updated CI scripts pointing to the new version of the docker images.

**Comment by prasad-alatkar** [comment]
> the Dockerfile update implies a version change. can you please split the dockerfile update into a separate PR (and include an update to integrations/docker/images/chip-build/version)?
> 
> The Dockerfile PR would need to land first, and a docker.io push will be necessary. Then this PR would have updated CI scripts pointing to the new version of the docker images.

Hi @rwalker-apple Dockerfile change is unnecessary one, I have made relevant change in my code so it negates the necessity of updating esp-idf from `tags/v4.1` to `release/v4.1`.  Please feel free to remove integrations label from the PR.

**Comment by prasad-alatkar** [comment]
Here is the summary of the footprint consumed by Bluedroid vs NimBLE BLE stack (for wifi-echo example). 

| Memory footprint | Bluedroid (Bytes) | NimBLE (Bytes) | Savings with NimBLE (KB) |
| -------- | -------- | -------- | -------- |
| Static memory  | 1185018  |  922782  | > 262KB |
| Free Heap  | 112784  | 140028 | > 27 KB |

**Review by andy31415 — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
@prasad-alatkar could you resolve the merge conflicts please?

**Comment by prasad-alatkar** [comment]
> @prasad-alatkar could you resolve the merge conflicts please?

I have resolved the merge conflicts. Thanks.

**Comment by bzbarsky-apple** [comment]
@saurabhst @BroderickCarlin @jelderton @chrisdecenzo @hawk248 could you take a look please?

**Review by BroderickCarlin — APPROVED** [review]
